core/src/main/kotlin/fr/sncf/osrd/stdcm/preprocessing/implementation/BlockAvailability.kt

```diff
@@ -34,13 +36,109 @@ data class BlockAvailability(
         startOffset: Offset<Path>,
         endOffset: Offset<Path>,
         startTime: Double
+    ): BlockAvailabilityInterface.Availability {
+        val pathStartTime = startTime - infraExplorer.interpolateDepartureFromClamp(startOffset)
+        val endTime = infraExplorer.interpolateDepartureFromClamp(endOffset) + pathStartTime
+
+        // Check that the planned step timing data is respected and return the corresponding
+        // availability
+        val stepAvailability = getStepAvailability(infraExplorer, startOffset, endOffset, startTime)
+        if (stepAvailability !is BlockAvailabilityInterface.Available) return stepAvailability
+
+        // Check the conflicts on the given path and return the corresponding availability
+        val conflictAvailability =
+            getConflictAvailability(startOffset, infraExplorer, pathStartTime, startTime, endTime)
+
+        val maximumStepDelay = stepAvailability.maximumDelay
+        when (conflictAvailability) {
+            is BlockAvailabilityInterface.Unavailable -> {
+                // If the resource is unavailable for a duration longer than the maximum step delay,
+                // it means the step timing data will not be respected, this block is forever
+                // unavailable
+                if (conflictAvailability.duration > maximumStepDelay)
+                    return BlockAvailabilityInterface.Unavailable(
+                        Double.POSITIVE_INFINITY,
+                        Offset(0.meters)
+                    )
+                return conflictAvailability
+            }
+            // Both are available: take the most restrictive, i.e., the one with the smallest
+            // maximum delay
+            is BlockAvailabilityInterface.Available -> {
+                if (conflictAvailability.maximumDelay > maximumStepDelay) return stepAvailability
+                return conflictAvailability
+            }
+        }
+    }
+
+    private fun getStepAvailability(
+        infraExplorer: InfraExplorerWithEnvelope,
+        startOffset: Offset<Path>,
+        endOffset: Offset<Path>,
+        startTime: Double
+    ): BlockAvailabilityInterface.Availability {
+        var maximumDelay = 0.0
+        first@ for (step in plannedSteps) {
+            for (location in step.locations) {
+                if (location.edge == infraExplorer.getCurrentBlock()) {
```

Comment:
We shouldn't just check the current block, during post-processing we check the availability of the whole path